python/ray/data/tests/test_streaming_integration.py

```diff
@@ -42,6 +42,19 @@ def ref_bundles_to_list(bundles: List[RefBundle]) -> List[List[Any]]:
     return output
 
 
+def test_autoshutdown_dangling_executors(ray_start_10_cpus_shared):
+    from ray.data._internal.execution import streaming_executor
+
+    initial = streaming_executor._num_shutdown
+
+    for _ in range(5):
+        ds = ray.data.range(100)
+        next(ds.iter_batches(batch_size=None, prefetch_batches=0))
+
+    final = streaming_executor._num_shutdown - initial
+    assert final == 5
+
+
```

Comment:
what's the behavior if you do something like this

```python
ds_iter = ray.data.range(100).iter_batches()
next(ds_iter)
ds_iter_2 = ray.data.range(100).iter_batches()
next(ds_iter_2)

next(ds_iter)
```

does this raise an error?

Reply:
It raises StopIteration, but this is a good observation. Changed this to RuntimeError with a better error message.

Reply:
@woshiyyya, @gjoliver, @matthewdeng - just to double check, PyTorch lightning trainer never reuses the iterator right? Given the [trainer has sanity check to iterate some validation batches before training](https://lightning.ai/docs/pytorch/stable/common/trainer.html#num-sanity-val-steps), I want to make sure we do not break anything here.

Reply:
hmm, why we raised an error in Amog's example? Those 2 iterators are for 2 different datasets right? I think this case is more reasonable

```python
ds = ray.data.range(100)

ds_iter = ds.iter_batches()
next(ds_iter)
ds_iter_2 = ds.iter_batches()
next(ds_iter_2)

next(ds_iter) # Raise error
```

Reply:
So there can only be one iterator working at the same time right? I did see some batch prefetching logic in Lightning, it might cause an error.